distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Regex/Match_2.enso

```diff
@@ -53,15 +62,176 @@ type Match_2
+                 match = Examples.match
+                 match.groups default="UNMATCHED"
+    groups : Any -> Vector (Text | Any)
+    groups self default=Nothing =
+        group_numbers = 0.up_to self.pattern.group_count
+        group_numbers.map n->
+            group = self.group n
+            case group of
+                Nothing -> default
+                _ -> group
+
+    ## Gets a map containing the named capturing groups for the pattern,
+       replacing the value for groups that did not participate in the match with
+       `default`.
+
+       Arguments:
+       - default: The value to return for a given name when the group at that
+         index did not participate in the match.
+
+       Note that it is possible for a group to "not participate in the match",
+       for example with a disjunction. In the example below, the "(d)" group
+       does not participate -- it neither matches nor fails.
+
+            "ab((c)|(d))".find "abc"
+
+       In this case, the group id for "(d)", which is 3, is a valid group id and
+       (Pattern_2.lookup_group 3) will return 3. If the caller tries to get group 3,
+       Match_2.group will return the default value.
+
+       > Example
+         Get the map of all of the named groups in this match, replacing the
+         value for groups that didn't participate in the match with "UNMATCHED".
+
+            pattern = Regex_2.compile "(.. .. )(?<letters>.+)()??(?<empty>)??"
+            input = "aa ab abc a bc bcd"
+            match = pattern.match input
+            ## match.named_groups.keys.sort == ["empty", "letters"]
+    named_groups : Any -> Map Text (Text | Any)
+    named_groups self default=Nothing =
+        named_group_names = self.pattern.group_names
+        spans = named_group_names.map name-> self.group name default=default
+        Map.from_vector (named_group_names.zip spans)
+
+    ## Gets the grapheme span matched by the group with the provided index, or
+       a default value if the group did not participate in the match.
+       If the identifier is invalid then `if_missing` is returned.
+
+       Arguments:
+       - id: The integer index or name of that group.
+       - if_missing: The value to return if the index is out of bounds.
+    get : Integer -> Any -> Any
+    get self index ~if_missing=Nothing =
+        self.group index . catch No_Such_Group (_-> if_missing)
+
+    ## Gets the grapheme span matched by the group with the provided index, or
+       a default value if the group did not participate in the match.
+       If the identifier is invalid then Index_Out_Of_Bounds is thrown.
+
+       Arguments:
+       - id: The integer index or name of that group.
+       - if_missing: The value to return if the index is out of bounds.
+    at : Integer -> Any ! Index_Out_Of_Bounds
+    at self index =
+        self.group index . catch No_Such_Group _->
+            Error.throw (Index_Out_Of_Bounds.Error index self.pattern.group_count)
```

Comment:
standard trick for `at` is to just call through `get` with `Error.throw (Index_Out_Of_Bounds.Error index self.pattern.group_count)` as `if_missing`.